### PR TITLE
Move wp-desktop files to wp-calypso monorepo

### PR DESCRIPTION
**Pull request by jsnajdr**
Moves files from `wp-desktop` repo to the Calypso monorepo and makes the build work.

**TODO:**
- better description of what I needed to do
- split the patch into multiple commits
- add CircleCI task to build and test the desktop app
- try to make `wp-desktop` development mode work
- convert the Docker environment for Windows debugging
- migrate the e2e utilities for desktop app, make the e2e suite green

**How to test:**
1. Run `yarn run build-desktop`
2. Run and test the app in `desktop/release/mac/WordPress.com.app`
3. While in the `desktop/` directory, run the app with `./node_modules/.bin/electron .` and test it

### CODE REVIEW

**Review by wp-desktop — DISMISSED** [review]
WordPress Desktop CI Failure (ci/wp-desktop): @jsnajdr please re-try this workflow ("Rerun Workflow from Failed") and/or review this PR for breaking changes. Please also ensure this branch is rebased off latest Calypso master.

**Comment by nsakaimbo** [comment]
Thanks for working on this, @jsnajdr! You've made impressive progress so far, and looks like you've tried to be as comprehensive as possible when porting things over. I'm still making my way through the changes, but a couple quick questions/clarifications:

1. Looks like the Makefile is being dropped in its entirety, and a lot of logic is being converted to JS, which is fine. However I'm not sure if the Docker utilities are captured, which are necessary to make the application debuggable on Windows. (I'm not sure if we're OK to drop this in lieu ofCalypso's existing Dockerfile. Will need to test, but it's not a deal breaker and we can always re-introduce if the workflow here is hampered.)
1. Similarly, utilities and scripts for end-to-end tests ([link](https://github.com/Automattic/wp-desktop/tree/develop/test)) look like they've been dropped as well (or I just can't locate them?). These utilities do things like capture video and logs of each test run, as well as make it easier to invoke tests locally for maintenance and debugging. All of this functionality is critical, and we should ensure to carry it over prior to landing the migration.
1. The existing root-level README should be included as well, as there are some important notes on application notarization, as well as building and packaging the app.
1. I'm assuming changes to the Circle config should be part of this PR as well?
1. We should double-check to update Calypso's .gitignore for items ignored in the Desktop's ignore file.

**Comment by jsnajdr** [comment]
Thanks for reviewing the PR, @nsakaimbo :+1: And for pointing out the missing parts. I added the nontrivial ones to the TODO list in the PR description. None of them were removed intentionally.

File: package.json
Comment (nsakaimbo) [inline package.json:70]: We'll want to ensure that additional arguments can be passed to the `build-desktop:app` script in order to override electron-builder's default behavior. One important optimization we make is building only the application executables if the desktop's config is unchanged (by passing the `"dir"` argument to electron builder). By adding this optimization, artifact build times in CI take ~5 minutes instead of ~18 minutes _after_ Calypso is built (see [PR](https://github.com/Automattic/wp-desktop/pull/869)).

Comment (nsakaimbo) [reply package.json:70]: I _think_ something like `electron-builder ${ELECTRON_BUILDER_ARGS:-""} build --publish never` should work (i.e. we want a default, empty value of `ELECTRON_BUILDER_ARGS` that we can override as necessary).

Comment (jsnajdr) [reply package.json]: OK, I added the `ELECTRON_BUILDER_ARGS` option to the `build-desktop:app` task. Now I run the build like this:
```
NODE_ENV=production ELECTRON_BUILDER_ARGS=-c.mac.target=dir yarn run build-desktop
```
And the app build is indeed much faster! The `WordPress.com.app` directory is built just in 3 seconds, while the whole ZIP + DMG task takes 46 seconds.

Comment (jsnajdr) [reply package.json]: By the way, I noticed that the local build is incredibly slow if the `node_modules` directory contains unneeded packages. If any of them is a native one (i.e., has some `node-gyp` bindings that require C compilation), `electron-builder` rebuilds them for the target platform, and that's both slow and useless.

That's why the desktop build needs to happen inside a dedicated `desktop/` directory, and the `package.json` file there needs to be careful about what NPM modules it has in `dependencies`.

**Comment by jsnajdr** [comment]
In the two most recent commits, I tried to reduce the size of stuff we bundle and ship with the desktop app.

First, I eliminated the bundled `node_modules` folder, telling webpack to include everything needed into the now self-contained server bundle.

Second, I turned on minification for both server and client for the desktop app.

**Results:**
The currently released WordPress.com app has DMG download size **100.2 MB**. The Calypso directories (`Contents/Resources/app`) are:
```
 70M	./node_modules
384K	./public_desktop
5.8M	./build
 72M	./calypso
148M	.
```
Here, `build` contains the server code (run by the "main" process, in Node.js), and the `node_modules` folder is there to satisfy `require` calls from the server bundle.

`calypso` is the client bundle, with all the JS assets and also images etc.

After eliminating `node_modules`, the DMG download size is **82.2 MB** The directories break down as follows:
```
 12K	./config
384K	./public_desktop
 72M	./public
 12M	./build
 85M	.
```
The `node_modules` folder is gone, and the `build` folder got 6 MB bigger. A good price to pay for eliminating 70 MB of `node_modules`.

The `public` folder is the same thing as `calypso`. It got renamed during the merge of `wp-desktop` to `wp-calypso`. At this moment, the size is the same.

After enabling minification for both server and client, the DMG download size is **77.3 MB** The directories look like this:
```
 12K	./config
384K	./public_desktop
 51M	./public
5.4M	./build
 57M	.
```
The `build` and `public` folder got significantly smaller.

This directory structure contains _everything_ that's needed to run Calypso, and it's almost identical to what runs in our Docker containers that serve the wordpress.com production.

This is how big (or small) our Docker images could be if they didn't contain the complete Git repository and the full `node_modules` folder with all the dev tooling. Could be interesting info for @griffbrad @sirreal @scinos.

**Comment by jsnajdr** [comment]
@nsakaimbo @belcherj This PR is currently in a state where local `yarn run build-desktop` produces a very nice desktop app that actually works.

How do you feel about merging the PR and then making the CI and e2e jobs work, before finally sunsetting the `wp-desktop` repo? And regularly syncing the changes in `wp-desktop/develop` into `wp-calypso/master` until then.

When I look at the CI and e2e work that needs to be done, it looks daunting and totally overwhelming. I prefer to work in smaller increments.

File: package.json
Comment (nsakaimbo) [reply package.json]: > ... electron-builder rebuilds them for the target platform, and that's both slow and useless

This step may add some overhead but is necessary as native modules require compilation for their target platform.

**Comment by nsakaimbo** [comment]
@jsnajdr Amazing work! 🎉 This is fantastic - excited to hear about all the optimizations you were able to find so far.

Regarding CI - I would be a little nervous about landing this without any CI whatsoever. I just touched base with @belcherj and we're switching up our schedule so I'll be able to dedicate my time to helping on this front. At the very least, I think we should have a very minimal CI job (i.e. Mac only) that is part of this prior to landing, and we can augment as we go. I believe I can have something ready-to-go within a few days (targeting early next week), at which point we can merge this.

**Comment by nsakaimbo** [comment]
Now that I think about it, I'm wondering if we can merge this branch prior to CI/e2e additions. My main concern is this may somehow interfere with CI status for the existing Calypso/wp-desktop setup. I could be wrong but it seems like it actually shouldn't, yeah? Folks can continue with their work and the existing desktop CI should continue to work, until we actually swap it out. If my understanding here is correct then we probably can proceed, but correct me if I'm wrong. If this will start failing existing desktop CI, we should probably hold off until we have a minimal implementation ready.

**Comment by scinos** [comment]
A few general notes:

- It is hard to understand what changes (if any) have been made to `wp-desktop` files, as all they appear as new here. What do you think about having `wp-desktop` as a git submodule initially while we set up the CI, scripts, etc?

- What's the reason for having some files in `client/desktop` and some in `desktop`?

- Could we make `client/desktop` its own package with its own dependencies? I think `client/package.json` is already too bloated and we should try to leverage the monorepo and workspaces and have many individual packages instead of a monolithic one.

- Regarding bundling and minifying server code: In my opinion it is not a good way to reduce server footprint. Node.js doesn't get any tangible benefit from having bundled or minified files, and it makes some things (like live debugging or interpreting stack traces) harder. If we want to reduce the size of the server image I'd start with splitting it from `client` to its own separate package, so it can have its own `node_modules`

**Comment by jsnajdr** [comment]
> It is hard to understand what changes (if any) have been made to `wp-desktop` files, as all they appear as new here.

I try to make that clearer by creating separate commits. There is a commit that imports the files verbatim, and then a different commit that makes modifications. The only modification I need to do there is to change the import paths by adding a `desktop/` prefix.

> What do you think about having `wp-desktop` as a git submodule initially while we set up the CI, scripts, etc?

It looks like a lot of extra work coordinating all the folders and import paths, and I don't think the benefit is big enough to justify it.

> What's the reason for having some files in `client/desktop` and some in `desktop`?

`client/desktop` contains modules that are part of the browser or server bundle that's shipped with the desktop app.

The browser bundle contains IPC event listeners that listen for commands sent by the main process , and also sends some IPC events back.

The server bundle has a custom boot code that load a lot of Electron-specific functionality, and then starts the mostly standard express.js server with handlers like _server/pages/index.js_.

The `desktop/` package/folder doesn't contain any app code, but rather provides an isolated environment for the Electron builder. Which runs with this directory as _cwd_. There are no packages in _desktop/node_modules_, it has _package.json_ that's fit for bundling inside the Electron app...

> Could we make `client/desktop` its own package with its own dependencies?

Yes, but not in this PR 🙂 This one is already complex enough and I have some difficulties making it hold together.

> Regarding bundling and minifying server code: In my opinion it is not a good way to reduce server footprint.

I can disable the minification. It's not needed at all to make the monorepo migration work.

It's also possible to set up the Terser minifier so that it, for example, only eliminates unused code (i.e., performs tree shaking) and doesn't transform the code that remains.

> If we want to reduce the size of the server image I'd start with splitting it from client to its own separate package

I agree with this direction, but I see it as a long term task. To make `client` and `server` separate packages, any code that their share needs to be a package, too. Currently, when they live in a common package, we can use simple relative imports (e.g., `../lib/user`) everywhere.

Packagifying all the shared dependencies is also a challenge for the proposed `desktop` package.

**Comment by jsnajdr** [comment]
> I would be a little nervous about landing this without any CI whatsoever

If you don't use it, the new desktop code just lies around in the repo and doesn't influence any existing stuff.

I only had to update the patch today so that the existing `wp-desktop` app can still build Calypso as a submodule. Some improvements will need to be postponed, like using the `evergreen` build instead of `fallback`, but we can merge them again later after we can break the old Desktop build.

The `wp-desktop` build and its e2e tests are now all green ✅

**Comment by nsakaimbo** [comment]
> The wp-desktop build and its e2e tests are now all green

Amazing! I was worried that we wouldn't have the option of gradually transitioning things over, but if we can continue to use calypso as a submodule in the `wp-desktop` repo while we migrate things gradually, that will provide a great deal of flexibility.

What sha of the wp-desktop repo is reflected in this branch? (Just want to keep track for my own reference so we be methodical about updating files as we go.)

File: package.json
Comment (jsnajdr) [reply package.json]: > That's why the desktop build needs to happen inside a dedicated `desktop/` directory, and the `package.json` file there needs to be careful about what NPM modules it has in `dependencies`.

The app doesn't need any native modules though. In our case, they are compiled only because `electron-builder` found them in the top-level `package.json` and thought that the app needs them. It didn't realize they are dev dependencies of the monorepo. The `electron-builder` doesn't recognize nuances like this.

**Comment by nsakaimbo** [comment]
I just tried taking this for a spin and got some build then runtime error (still trying to debug).

1. check out `import/wp-desktop` branch
1. `yarn install --frozen-lockfile` (ensure Node 12.18.0 is being used)
1. run `NODE_ENV=production ELECTRON_BUILDER_ARGS=-c.mac.target=dir yarn run build-desktop`

The first time I attempted the build, I got a node-gyp error: 

```
warning Error running install script for optional dependency: "/...chokidar/node_modules/fsevents, /.../jest-environment-enzyme/node_modules/fsevents, /.../newspack-blocks/node_modules/fsevents: Command failed.

gyp ERR! stack Error: Python executable \"/usr/bin/true\" is v, which is not supported by gyp
```

I double-checked and the version of Python installed on my machine is v2.7.16. I'm guessing this is partially what you were referring to, @jsnajdr (i.e. that electron-builder will attempt to rebuild all native modules it finds)?

I tried to build the app again and this time the app built to `desktop/release/mac` (could be that node-gyp is relying on some cached state at this point so the first error didn't repeat). However, double-clicking the executable opened a corrupted app with `NOT FOUND` displayed.

<p align="center">
Screen Shot 2020-06-22 at 3 27 45 PM
</p>

The Node process seems to be running - not sure why the renderer process is erroring out. From the application logs (`Help -> Get Application Logs`):

```
[2020-06-22 15:35:05.609] [desktop:boot] [info] Settings:{"notification-badge":true,"notification-bounce":true,"release-channel":"beta","proxy-type":"","proxy-url":"socks5://localhost","proxy-port":"8080","proxy-pac":"","spellcheck-enabled":true,"wpDragAndDropDisabled":true,"window":{"x":501,"y":327,"width":1873,"height":913},"last-location":"/me/security/two-step"}
[2020-06-22 15:35:05.609] [desktop:boot] [info] Proxy: none
[2020-06-22 15:35:05.624] [desktop:index] [info] Waiting for app window to load
[2020-06-22 15:35:05.878] [desktop:index] [info] Starting window handlers
[2020-06-22 15:35:05.621] [desktop:updater] [info] Skipping Update – no configuration
[2020-06-22 15:35:05.779] [desktop:server] [info] Starting server
[2020-06-22 15:35:05.805] [desktop:server] [info] Server created, binding to 41050
[2020-06-22 15:35:05.808] [desktop:server] [info] Server started, passing back to app
[2020-06-22 15:35:05.731] [desktop:runapp] [info] No other instances, waiting for app ready
[2020-06-22 15:35:05.768] [desktop:runapp] [info] App is ready, starting server
[2020-06-22 15:35:05.809] [desktop:runapp] [info] Loading app (http://127.0.0.1:41050/me/security/two-step) in mainWindow
[2020-06-22 15:35:05.809] [desktop:settings] [info] Get setting for last-location = /me/security/two-step
[2020-06-22 15:35:05.810] [desktop:settings] [info] Updated settings for group 'window': {"x":501,"y":327,"width":1873,"height":913}
[2020-06-22 15:35:05.810] [desktop:settings] [info] Get setting for spellcheck-enabled = true
```

Nothing in the boot sequence seems unusual.

However, starting the built app from the command line seems to indicate an issue with resolving `assets-fallback.json`:

```
> ./desktop/release/mac/WordPress.com.app/Contents/MacOS/WordPress.com

Error: ENOENT: no such file or directory, open 'client/server/bundler/assets-fallback.json'
    at Object.openSync (fs.js:440:3)
    at Object.func (electron/js2c/asar.js:140:31)
```

**Comment by nsakaimbo** [comment]
> Packagifying all the shared dependencies is also a challenge for the proposed desktop package.

After playing around with the progress so far, I'm wondering if we would be better off separating the desktop's dependencies into it's own package.json. There might be some overlap between the two, but likely minimal. The benefit in keeping the two separate is avoiding node-gyp errors/unnecessary compilation, as well as making it conceptually easier for developers to grok what dependencies are required specifically for the desktop app. Is there a strong reason _not_ to do it this way?

I would imagine we could keep the rest as-is, so the workflow would be something like `cd into desktop && yarn install && cd to calypso && yarn build desktop-app`. An extra step to cd into the desktop folder and install dependencies, but I think overall a lot cleaner to manage and reason about.

**Comment by jsnajdr** [comment]
> The first time I attempted the build, I got a node-gyp error:

@nsakaimbo This is not happening on my machine 🙂 Few more details could help to reproduce and/or solve this:
- which build step exactly tries to do this `node-gyp` build? The `lerna run prepare` step? Or is it the `electron-builder` inside `build-desktop:app`? The full build log (of the `yarn run build-desktop`) command would be useful.
- what are the contents of your `desktop/node_modules` folder? It should be almost empty, with only the `.bin/electron-builder` executable.

> However, double-clicking the executable opened a corrupted app with NOT FOUND displayed.

That was a bug, and I fixed it today. The `client/server/bundler/assets-fallback.json` file was not being copied to the app bundle. Please verify with the latest version of the `branch`.

> After playing around with the progress so far, I'm wondering if we would be better off separating the desktop's dependencies into it's own `package.json`.

This branch kind of already does that, by creating the `desktop/` directory, with its own `package.json`, and running the `electron-builder` there. For the sake of simplicity, I eliminated all the dependencies from there, so that the bundled app doesn't need to ship any `node_modules` folder.

**Comment by nsakaimbo** [comment]
> which build step exactly tries to do this node-gyp build?

To rule out conflicts/errors with cached state, I re-cloned the entire wp-calypso repository into a brand new folder on my machine. Running `yarn install --frozen-lockfile` produced the same node-gyp error. I can confirm this only seems to happen when the root `node_modules` folder is empty - repeating the same `yarn install --frozen-lockfile` command will not result in the error. However this will still be a concern (more so in CI).

I found that to reproduce, clear out the root node_modules folder (`rm -rf node_modules`) and run `yarn install --frozen-lockfile` again. The error happens during the `Building fresh packages...` step, _before_ `$ check-npm-client && yarn run build-packages`.

Inspecting the `desktop/node_modules` folder:

```
.bin
├── electron -> ../../../node_modules/electron/cli.js
├── electron-builder -> ../../../node_modules/electron-builder/out/cli/cli.js
└── install-app-deps -> ../../../node_modules/electron-builder/out/cli/install-app-deps.js
```
[yarn-install-node-gyp-error.log](https://github.com/Automattic/wp-calypso/files/4820219/yarn-install-node-gyp-error.log)

I'm attaching the full output from the yarn install command.

**Comment by nsakaimbo** [comment]
Tested SHA b2285e8813349c884bd1b057e5372b33b1b2859a and now the app isn't building at all. 😕

In a freshly-cloned wp-calypso folder:

1. Install node modules `yarn install --frozen-lockfile`
1. `NODE_ENV=production ELECTRON_BUILDER_ARGS=-c.mac.target=dir yarn run build-desktop`

```
$ check-npm-client && run-s build-desktop:config build-desktop:server build-desktop:client build-desktop:app
$ check-npm-client && node desktop/bin/build-desktop-config.js
$ check-npm-client && BROWSERSLIST_ENV=server CALYPSO_ENV=desktop webpack --config client/webpack.config.desktop.js --display errors-only
$ check-npm-client && CALYPSO_ENV=desktop DESKTOP_MONOREPO=true yarn run build-client-fallback
$ check-npm-client && BROWSERSLIST_ENV=defaults yarn run build-client-do
$ check-npm-client && node ${NODE_ARGS:---max_old_space_size=8192} ./node_modules/webpack/bin/webpack.js --config client/webpack.config.js --display errors-only && yarn run build-languages-if-enabled
$ check-npm-client && node -e "(process.env.BUILD_TRANSLATION_CHUNKS === 'true' || process.env.ENABLE_FEATURES === 'use-translation-chunks') && process.exit(1)" || yarn run build-languages
$ check-npm-client && node -e "process.env.DESKTOP_MONOREPO === 'true' && process.exit(1)" && yarn run validate-fallback-es5
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: "build-desktop:client" exited with 1.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
It may help to attempt to reproduce this by installing + building from a clean location as well.

File: package.json
Comment (nsakaimbo) [inline package.json:66]: There should be an additional check to ensure that `secrets.json` exists and is valid before proceeding to build the app, as with the original Makefile logic ([link](https://github.com/Automattic/wp-desktop/blob/develop/Makefile#L134-L145)) to mitigate the risk of an invalid app being built either locally or in CI.

**Comment by nsakaimbo** [comment]
> For the sake of simplicity, I eliminated all the dependencies from [desktop package.json], so that the bundled app doesn't need to ship any node_modules folder.

This is great, except I just realized that lifting all (or close to all) of the desktop's dependencies to Calypso means we can no longer smartly optimize build times in CI. The way we do this is by checking whether or not the desktop's dependencies or configuration have changed. If not, we proceed with a partial build (where we set `ELECTRON_BUILDER_ARGS=dir`). [Here](https://github.com/Automattic/wp-desktop/blob/2ae11911eb941f15fbf197ce887e0430e10e55de/.circleci/config.yml#L426-L428) is an example of where that logic lives today:

```
 # Build all artifacts only when project config changes.
 # Otherwise only build application executable required for end-to-end testing.
 ! git diff --name-only origin/develop...HEAD | grep -E -q 'package.json|yarn.lock' && ELECTRON_BUILDER_ARGS='-c.mac.target=dir'

make package ELECTRON_BUILDER_ARGS=$ELECTRON_BUILDER_ARGS
```

We have to build all artifacts when the desktop's config changes because that is when the risk is highest for breaking changes being introduced to installers and other artifacts. To emphasize, the build time differential here is drastic - less than ~5 minutes (including e2e tests) vs. ~18 minutes. The alternative is to condition this check on Calypso's yarn.lock instead - hopefully this would change infrequently enough that the occasional ~18-minute CI completion time is acceptable (I think this is a reasonable assumption but correct me if I'm wrong).

**Comment by jsnajdr** [comment]
> I seem to have resolved this by explicitly setting the python version in my yarn config

Your error log contains the following:
```
gyp ERR! configure error 
gyp ERR! stack Error: Python executable \"/usr/bin/true\" is v, which is not supported by gyp.
gyp ERR! stack You can pass the --python switch to point to Python >= v2.5.0 & < 3.0.0.
gyp ERR! stack     at PythonFinder.failPythonVersion (/Users/nsakaimbo/workspace/test/wp-calypso/node_modules/node-gyp/lib/configure.js:493:19)
```
Somehow, `/usr/bin/true` is used as the Python executable when trying to detect the Python version, and that's wrong. It should be something like `python`, `python2` or `python2.7`

That can happen when running `node-gyp --python=true` or `PYTHON=true node-gyp` or `yarn --python=true` or setting `yarn config set python true`. In short, something in your local environment sets the `python` variable to an invalid value. Yarn config, `.npmrc`, something else... Can you try to find it?

If you unset the Yarn config variable with `yarn config delete python`, does the "building fresh packages" step start failing again?

**Comment by jsnajdr** [comment]
> Tested SHA b2285e8 and now the app isn't building at all. 😕

I see, the condition in `postbuild-client-fallback` was wrong, making the task fail instead of silently skipping it. Fixed in latest version of the branch.

**Comment by jsnajdr** [comment]
> What sha of the wp-desktop repo is reflected in this branch?

It's https://github.com/Automattic/wp-desktop/commit/e299749600acd32606ac9ea28c6a5065a0c5b98c:
```
e299749 2020-06-08 22:49:27 +0200 nsakaimbo@users.noreply.github.com ci: add Slack notifications for Calypso PRs (#903)
```

Since then, https://github.com/Automattic/wp-desktop/pull/928 was merged among others. It won't be a problem to pull in these changes, either in this branch, or as one of the followups.

**Comment by jsnajdr** [comment]
> This is great, except I just realized that lifting all (or close to all) of the desktop's dependencies to Calypso means we can no longer smartly optimize build times in CI.

I hope that after I eliminated the runtime `node_modules` dependencies and started bundling all packages into one bundle, this optimization won't be neccessary.

18 minutes sounds truly excessive, and I suspect it was happening because `electron-builder` was rebuilding native `devDependencies` like `fsevents` or `sharp`. That should no longer be happening.

A comment on https://github.com/Automattic/wp-desktop/pull/869 says that most of the time is indeed spent installing and building `node_modules`:

> The delta between platform packaging and the total runtime of each platform's build job is primarily driven by the node module installation step.

Building the packaged artifacts should involve simply zipping the _WordPress.com.app_ folder into a compressed file. It's still a nontrivial CPU-intensive task (on my machine, it takes 40 seconds with CPU 100% utilized by the `7za` process), but I don't see how it could take 18 minutes.

I propose to set up the CI tasks and see how fast they are. Then we'll see if further optimizations are needed.

File: package.json
Comment (jsnajdr) [reply package.json:66]: Added the check in the latest commit. It's now in the script that runs during the `build-desktop:config` task.

**Comment by nsakaimbo** [comment]
> A comment on Automattic/wp-desktop#869 says that most of the time is indeed spent installing and building node_modules:

That intent of that statement was to clarify that the remaining delta - _after_ the first optimization pass - is driven by the node module installation step, which by itself takes a relatively trivial 1-2 minutes.

I spent a lot of time profiling build times when I did the initial work, and node module installation was not the main bottleneck to begin with. The actual artifact build times (specifically generating the embedded blockmap on Mac) by itself can take anywhere between 8-10 minutes. We shouldn't be generating these assets with every build.

> I propose to set up the CI tasks and see how fast they are. Then we'll see if further optimizations are needed.

Sounds reasonable. Once CI we should have a clearer picture and we can re-evaluate.

File: .circleci/config.yml
Comment (nsakaimbo) [inline .circleci/config.yml:630]: Building the desktop client + server source can take as long ~10 minutes ([link](https://app.circleci.com/pipelines/github/Automattic/wp-calypso/65123/workflows/69f62ec9-7575-4f8c-a4f2-1d734f48d4f5/jobs/787903)). 

As with the wp-desktop build today, we try to optimize this by building the source once, which is then sent to a downstream job for each platform to compile the final application bundle/binary.

File: .circleci/config.yml
Comment (nsakaimbo) [inline .circleci/config.yml:703]: Temporarily commenting out to speed up the feedback loop while we're iterating on this.

**Comment by nsakaimbo** [comment]
Interesting point of comparison: within each platform's build job (taking Mac as an example), the node module installation step now takes ~4.5 minutes as the Calypso and Desktop dependencies are combined: [link](https://app.circleci.com/pipelines/github/Automattic/wp-calypso/65123/workflows/69f62ec9-7575-4f8c-a4f2-1d734f48d4f5/jobs/787918). 

In contrast, this step was taking < 1m until now: [link](https://app.circleci.com/pipelines/github/Automattic/wp-desktop/16875/workflows/cbb1cd3a-45e1-4203-b119-482ad6e61817/jobs/65359/steps).

Because of this overhead, the Mac job that would take 3-5 minutes to complete will now take roughly 7-9 minutes instead (factoring e2e tests). A concern here is it the time added to developer workflow when running (or re-running) CI.

Perhaps there's some way to optimize this, but none that I've found so far. AFAIK the node module cache is not compatible across platforms (so we can't use a module cache that was generated in an upstream Linux job for downstream Mac or Windows jobs, for example).

**Comment by nsakaimbo** [comment]
When building the JS source, the occasional EPIPE error is occurring: [link](https://app.circleci.com/pipelines/github/Automattic/wp-calypso/65123/workflows/742de093-c72b-454a-abc1-8a5300da6c51/jobs/787889/steps)

I believe this is the same issue as reported in #43533 so isn't unique to this PR specifically.

**Comment by jsnajdr** [comment]
> the node module installation step now takes ~4.5 minutes as the Calypso and Desktop dependencies are combined

We added the `electron` and `electron-builder` packages that contain many large binary files. These are not needed by any other task but desktop build, and maybe we could also cache their content better. To avoid large downloads or C compilations on `yarn install`.

A possible solution is to further isolate the `desktop/` directory from the rest of the monorepo. Move it out of the Yarn/Lerna workspace structure, and run a separate `yarn install` with a separate `yarn.lock` in `desktop/`.

None of these large packages is needed to build the sources into the client/server JS bundles. They are only needed to create the packaged Electron app.

**Review by nsakaimbo — APPROVED** [review]
I attempted to augment @jsnajdr's work by getting CI to work in this branch, but more work will be required to debug. In the mean time, this work does not interfere with the existing desktop build and CI across the desktop canary bridge.

The work in this branch looks good, and I confirmed the application is building successfully locally. Given the additional complexity of integrating CI, I think we're fine to proceed with getting this merged for now and staggering CI integration to a different PR. 

Thanks for the awesome effort and feel free to merge once all CI is ✅ .

**Comment by nsakaimbo** [comment]
> A possible solution is to further isolate the desktop/ directory from the rest of the monorepo. Move it out of the Yarn/Lerna workspace structure, and run a separate yarn install with a separate yarn.lock in desktop/

We should do absolutely do this as there is another limitation of the current approach. Calypso packages cannot be built on Windows. The node module installation step automatically builds Calypso's packages, so if we attempt to build the app for Windows CI, the build will bomb. 😕 So we may have no choice but to keep their dependencies separate. 

With the existing submodule setup, the way it works today is:

1. Desktop and Calypso source are built in Linux
2. Built source is sent to downstream Mac, Windows and Linux jobs ([here](https://app.circleci.com/pipelines/github/Automattic/wp-desktop/16875/workflows/cbb1cd3a-45e1-4203-b119-482ad6e61817/jobs/65361/steps) is an example Windows job - no issues here as Windows is not trying to compile any of the Calypso source or packages)
3. The downstream platform-specific jobs install node modules, but only for the desktop (not Calypso)
4. Pre-built source is compiled with Electron binaries to produce each platform's final build

I'm currently working on CI for the monorepo migration (#43709). Isolation of the desktop's dependencies would be required to get Windows working as well.

**Review by nsakaimbo — DISMISSED** [review]
I was doing another pass through the project's dependencies, and I see at least a few existing desktop dependencies are missing from both the root-level package.json and `desktop.package.json`. Some key dependencies will actually break functionality if missing:

"semver": "5.6.0" (used in "About" window and for the auto-updater)
"archiver": "3.1.1" (used for logging)
"electron-fetch": "1.2.1" (auto-updater)
"electron-updater": "4.2.5" (auto-updater)
"js-yaml": "3.12.0" (manual updater)
"lodash.debounce": "4.0.8"
"make-dir": "3.0.2" (application logging)
"portscanner": "1.2.0" (used in desktop/server logic)
"winston": "3.2.1" (application logging)

I confirmed that the desktop source actually imports the above-mentioned libs in various places. Could you please ensure these deps are accounted for before landing this? Might be worth doing another manual pass through the existing dependency list to double-check that I didn't miss anything.

**Comment by jsnajdr** [comment]
> I see at least a few existing desktop dependencies are missing from both the root-level `package.json` and `desktop/package.json`.

These dependencies you mention were correctly added to `client/package.json`. They are required by the server (main) and client (renderer) bundles and therefore belong to `client/`.

The root `package.json` contains the common devDependencies (i.e., the tooling), and `desktop/package.json` contains only the Electron tooling needed to build the app.

**Comment by jsnajdr** [comment]
In the last 2 commits, I pulled the `desktop/` directory out of the Yarn workspace structure. That creates a separate modules root in `desktop/` with its own `yarn.lock`.

If we didn't do that, the root `yarn install` would always install approx 300MB of binaries in the `electron` and `app-builder-bin` packages. However, we need these only when building the actual Electron app, never for anything else.

**Comment by jsnajdr** [comment]
I verified that the failing e2e tests are not related: the same failure (publicizing a post Twitter) happens in master, too.

Going to merge now: I believe that the concerns that led @nsakaimbo to "request changes" are addressed. And I want to do the merge near the beginning of my working hours.

**Comment by nsakaimbo** [comment]
> These dependencies you mention were correctly added to client/package.json. They are required by the server (main) and client (renderer) bundles and therefore belong to client/.

Apologies - I totally missed this! 🤦 Glad to hear everything is accounted for.

**Pull request merged**
